### ISSUE TITLE
Easy way to get Job

### ISSUE THREAD

**Issue by sagivf** [grantcarthew/node-rethinkdb-job-queue#47]
I am adding Jobs based on an existing RethinkDB table.
I would like to "Determine job uniqueness" but using `findJob` uses a filter which can be a little slow and resource consuming.

It looks like the jobs are created based on a UUID id.

If I could, I would give the job the same id as the original record in the other table and could use `getJob` which I assume fetches by the primary key.

Any advice? Is it problematic in your opinion to give a job your own ID?

**Comment by grantcarthew** [comment]
Hi @sagivf.

The Job Ids are generated within the constructor for the Job using the UUID package on NPM.
If you want to set the Job Id to a known existing UUID for the job, just do it before you add it to the queue.

```js

let job = Queue.createJob()
job.id = 'your custom uuid'

```

I see no reason that this would cause problems.

I haven't tried non-UUID ids. I don't know how that would work. It would require a little research and testing if you wanted to try it out.

**Comment by sagivf** [comment]
Thanks @grantcarthew 
Appreciate the quick response.

I tried it out and I get an error saying - "The job id is invalid".
I'm using the id generated by rethinkdb so I think it should be a valid uuid. I think it's being validated here - 
https://github.com/grantcarthew/node-rethinkdb-job-queue/blob/ba7c971d8c5626005b74ea417138e2c27b17819a/src/job-parse.js#L31

Perhaps the node `uuid` module is not compatible with rethinkdb's one.
I will try to investigate but if you have more knowledge let me know.

Thanks again.

**Comment by grantcarthew** [comment]
The validation code is testing for a [version 4 UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier). There are other versions.

Go into your `node_modules` directory to the `rethinkdb-job-queue` package, find the line of code you linked above within the `dist` directory and comment it out. Then try it out again.

If it works I think it is reasonable to publish a new version without that test. Your use case is quite a good one and should be supported.

**Comment by sagivf** [comment]
I've tried it out.

There are two places need changing - commenting out that line and changing these lines - 
https://github.com/grantcarthew/node-rethinkdb-job-queue/blob/ba7c971d8c5626005b74ea417138e2c27b17819a/src/job-parse.js#L11,
I'm not exactly sure why the job can sometimes be an ID. we could change that to `typeof string` or something of the sort. 

There is one pitfall here that I ran into and others might as well.
If you try to insert a job after completion, it will throw an error "duplicate id" because the old item isn't deleted yet, event when using `removeFinishedJobs: true`. Using a timeout works or maybe using `repeat: true` instead (haven't tested that out - should be tested).
What do you think? I can open a pull request but I think this should be more explicit like adding a `customId` property to the config options  with proper warnings in the documentation.

**Comment by grantcarthew** [comment]
Ok, understood. Thanks.

There is another option. We could change the RethinkDB `insert` statement to include the option `conflict: replace`.

Ref: https://www.rethinkdb.com/api/javascript/insert/

This would allow adding jobs with the same Id by replacing them with the new data.

As you stated. It might be better to add that as a Queue option `enableCustomJobIds` or something like that. If that option is true, Id parsing is disabled and `conflict: replace` is enabled. Could also add a `Job.setId` method.

What do you think?

**Comment by sagivf** [comment]
- I would say that if there is a clear in indication that the job is complete - `conflict: replace` is a good solution. There is a `conflict: function(){}` by the way and perhaps the job status could be checked there.
If the job isn't completed I would say an error should be thrown.

- `enableCustomJobIds` sounds great, that way you have to opt in and there is no room for confusion.

- Personally, I think sending an id as a param to the job creation should be enough, any reason why not?

**Comment by sagivf** [comment]
Let me know if you want me to open a pull request

**Comment by grantcarthew** [comment]
Sometimes the obvious solution is easy to overlook. You're right, Passing Id as a parameter sounds like a KISS solution. If we pass the Id as a parameter we would need to change the `is.job` and Id parsing to prevent issues.

The `is.job` test within the Job constructor is quite important. It determines if the job is being created from fully populated data (typically from the database). It might be worth adding a few more property tests to it if we are not testing the Id.

I think this is a better way to go than changing the Id after job creation. However, there would still be the issue of Id conflicts if the job already exists. Rather than adding a Queue option, a second method argument to `Queue.addJob(job)` could be used to force a replace. Something like `Queue.addJob(job, clobber)`.

There are problems with this idea also:

*   The job would have to be tested to ensure it is not active at the time of adding a job with an existing Id.
*   There could be a race condition if you have the Queue removing jobs.

These are all issue that are irrelevant if you don't allow control of the Job Ids.

Maybe there is an alternative solution for your current problem @sagivf. What's the existing table you are talking about?

Once we have hammered out the details and if a change is needed I am happy if you are keen to work on it. Otherwise I can make the changes. Have a read of the [Contributing](https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Contributing) document prior to coding.

**Comment by grantcarthew** [comment]
@sagivf 
I have been thinking about this issue and am leaning towards not enabling custom job Ids because of the issues caused.

I don't know all of the details about your situation however I did think of a possible solution. Could you create a job queue that is used to sync the jobs? It's a little bit of double handling however it may fix your issue.

**Comment by sagivf** [comment]
Hey @grantcarthew.
Sorry for the delayed response, was a busy week...

Not sure I completely understand your last solution. But here are my recent thoughts:

- Looking up a job for status checking or existence is a pretty common use case, do you agree?
- Currently the are two ways to "find" a job 
   a) Save the generated id somewhere - this can be a hassle and not performant.
   b) Add a property to the job and use `queue.findJob` - is great but less performant.
- This isn't a critiacal feature as (b) is usually fine.
- If you do decide to attack this problem I think I have a solution:
a) add an option for a `name` property on jobs and an index on the `name` field. not sure if `name` needs to be configurable - I think not.
b) add a `fetch(name)` function that will find by index on name. (can return more then one - needs documentation to avoid confusion).
c) add a `contains(name)` function that will check if there is at least 1 result in `fetch(name)`.
I think this will help my problem and add more power to the projects api.

**Comment by grantcarthew** [comment]
Ok. I don't mind the idea of adding a `Job.name` property with an index. It could default to the same as the job Id. Then have the `fetch(name)` and `contains(name)` methods use the new `name` index.

Let's look at the API:

*   `Job.setName(string)` - could also use `job.name = string` however worth having this method.
*   `Queue.findJobByName(string)` - more descriptive than fetch and matches existing API?
*   The existence test has a number of options:
    *   `Queue.contains(string)`
    *   `Queue.includes(string)` - should it be `includes` like the other JS APIs such as `Array.includes()`?
    *   `Queue.includesJobName(string)` - again, more descriptive?
    *   `Queue.hasJobByName(string)` - move away from contains and includes and more descriptive?
    *   `Queue.jobExists(string)`
    *   `Queue.jobExistsByName(string)`

I wonder what other queues are using.

**Comment by grantcarthew** [comment]
I've been thinking on this further and I don't see the need for an existence method. It is trivial to test for a positive result from `Queue.findJobByName(string)`.

There may be a minor performance improvement by using the [contains RethinkDB API](https://rethinkdb.com/api/javascript/contains/). However without testing this is unknown. The only gain would be not having to return the job document(s).

**Comment by sagivf** [comment]
A few points:
1) Since it's a secondary index there can be any number of jobs with the same name name it should be `Queue.findJobsByName`
2) It's true the existence method is trivial but I think it still helps the user in keeping his code clean and readable.
3) I think JS uses `includes` cause the spec wanted to avoid conflicts with exiting `Array.prototype` contains overrides. But I don't have a clear opinion on which name.
3) Regarding naming, I don't mind each way but I think `fetch(name)` already indicated the fetch is by name.
4*) I'm pleased with what we have come with so far, but there is one more idea to consider -
we could hold an `id` and `_id` property. If the user doesn't supply an id, both are set to the same UUID. If he does `_id` is set as the primary key and self generated. and `id` is used for  `fetch/findJobById` with a secondary index. We could reverse that to make the implementation easier, but it seems less correct to me.

**Comment by grantcarthew** [comment]
1.   Most of the methods on the Queue return an array. I stuck with the singular naming because it felt right at the time.
2.   Existence test in or out is easy to do. I don't have a need for it however you do. If you think it is better to have we can add it.
3.   I like `has` I think. `Here is a queue. It has this job in it.` I sometimes forget which one to use, includes, contains, when working in JS then Python etc. If you read though the [Queue API](https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki), 'has' seems to fit in. That said, contains would be fine also.
4.   The `name` is a variable. User code could call it anything. In my experience being verbose with variable and method naming has never caused problems, being succinct does.
5.   I understand how you are thinking however calling it Id will cause people to think it is unique. Your first idea, name, was a good idea.

## Proposed API
| API        | Description       |
|----------|------------------|
| `Job.setName(string)` | Sets the name of the job. Defaults to `Job.id` |
| `Queue.findJobByName(string [, raw])` | Returns one or more Job objects or raw data |
| `Queue.hasJobByName(string)` | Returns `true` or `false` or a positive integer if more than one? |

## Proposed Work

*   Add `indexName` RethinkDB index.
*   Add `Job.setName(string)` Job method.
*   Add `Queue.findJobByName(string [, raw])` Queue method using the new index.
*   Add `Queue.hasJobByName(string)` Queue method using the [count](https://rethinkdb.com/api/javascript/count/) or [contains](https://rethinkdb.com/api/javascript/contains/) RethinkDB API. Depending on if we return a positive integer or boolean.
*   Test all the above changes.
*   Document the changes.

## Notes / Question

*   I don't want to enable users of the queue to be able to change the Ids. I have been watching the other queue packages for a while now and there are always issues raised because of the sorts of problems that arise due to the Ids. This does make the `rethinkdb-job-queue` package a little less flexible for the gain of stability.
*   Do you think returning a positive integer if there are multiple named jobs is a good idea for `hasJobByName`? I like the idea. I don't know what the performance differences between `contains` and `count` would be like though. It would make sense that `contains` would be faster making the positive integer not such a good idea.
*   All of these changes are not changing existing API so the version would be 2.2.0
*   I like these additions Sagiv. Thanks for the input. If you are keen to contribute, have a read of the [contributing](https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Contributing) wiki page. Following StandardJs and making improvements where you see them.

**Comment by grantcarthew** [comment]
Something just occurred to me. I think `Queue.hasJobByName` will need more arguments. Do you want to test for all jobs by the name, or jobs which are not completed?

With the `findJob` methods, this can be done by client code. Not so with an existence test.

Maybe `Queue.hasJobByName(name, includeCompletedJobs)` where `name` is a string and `includeCompletedJobs` is a `boolean` defaulting to `false`.

This makes me think the existence test should not be included in this feature.

**Comment by grantcarthew** [comment]
FYI, I'll have some time this week to work on this.

**Comment by sagivf** [comment]
Great, though theres currently no rush on my side.

Regarding contribution - I'm actually leaving my current job in a few weeks and want to do some open source work. I want to contribute to the web parts of open-rethinkDB and this to is a great project to contribute to as well! 
If you open a branch let me know and I will follow it for know.

kudos on the summarization of the issue, nicely done and 👍  on all the details! 
A few things:
- `Queue.findJobByName(string [, raw])` what is raw?
- Regarding the positive integer - I think `count` is o(n) and rethink has had some issues with it. not sure regarding `contains`. I would separate it into a it's own api tough I understand where your coming from.

One other thing:
If someone is interested in something like https://github.com/rschmukler/agenda and https://github.com/moudy/agenda-ui. How close is this project to it? I'm guessing a new project which uses `node-rethinkdb-job-queue` would be pretty easy, no?

**Comment by grantcarthew** [comment]
Hi @sagivf. I have finished #49 code and it is published as v2.2.0. I'm just working on the documentation updates. Once that is done I will work on this one.

*   [Raw is explained on the Queue.findJob document](https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Queue.findJob)
*   I am going to leave the existence check out for now. It can be added later.

Can you please expand on the `agenda` suggestion/question? I'm not really short what you are wanting to point out. RJQ has [repeat](https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Job.setRepeat) already but not scheduled times support.

**Comment by sagivf** [comment]
So I originally used `agenda` to manage my jobs and `agenda-ui` is a nice visual add on.
After switching from mongo to rethinkdb i tried https://github.com/iamkevla/agenda-rdb but it's under maintained and buggy. I switching to `rethinkdb-job-queue`and realized it is much richer and powerful.

So the way I see it, wrapping/adding some scheduling options would make this project more accessible. the repeat feature would help a lot here.

I agree It shouldn't be a part of this project. But I think combining a scheduling system with `rethinkdb-job-queue` along with some nice api's **might** be a goos idea.
Was wondering what where your thoughts...

**Comment by grantcarthew** [comment]
I'm keen to add useful features to the project without adding complexity.

It took me a long time to get [Queue.process](https://github.com/grantcarthew/node-rethinkdb-job-queue/blob/master/src/queue-process.js) to its current stable state. It is a complex mix of functions and references. If it wasn't for tests I would have never gotten it to this state.

With that in mind, anything that touches Queue.process may be add adding more complexity to an already complex file.

Adding scheduling will need some extra attention because the RethinkDB doesn't have an alerting system to raise events at specific times. This means you would need to create a schedule document which will need to be maintained in memory once the Queue object is initialized. I haven't put much thought into this yet.

So adding scheduled job processing is worth investigating for sure however it will need a good design.

With regards to the UI, this project could do with one for sure. @tomkaltz and @thomasmodeneis have both discussed a web front-end in #37. I'm not keen to work on this project myself because I have my personal projects to work on.

If you guys want to discuss the front-end we can open an new issue.

With regards to adding the features in this issue, they are already done in master apart from the existence check. I will publish the package today if I can.

Thanks for your input mate.

**Comment by grantcarthew** [comment]
I just opened #51 to discuss a UI.

**Comment by grantcarthew** [comment]
@sagivf This is all done now.

See the [Change Log](https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Change-Log).

I decided to go with `containsJobByName` rather than `has` because it is more common for existence testing.

After researching the best way to perform a `contains` query against the RethinkDB database, it turned out using `getAll` with the `name` index is the best option. So the `Queue.containsJobByName(string)` is simply syntax sugar over the `Queue.findJobByName()` method checking the returned array length.

I'll leave this open until the documentation is updated.

**Comment by grantcarthew** [comment]
Documentation is mostly done. Just finishing of the Complex Job document. Will revise it tomorrow.

I'll close this now.

Thanks again.

**Comment by sagivf** [comment]
Thanks @grantcarthew I will check it out soon and if any issue arise I will let you know.
